### ISSUE TITLE
Feature selection + hyperparameter tuning + model validation (nested-nested CV)?

### ISSUE THREAD

**Issue by tjpalanca** [mlr-org/mlr#1861]
There are some learners (e.g. Random Forests) where I have to do: (a) partition dataset into training and test, (b) select features within the training set via Wrapper approach, and (c) tune hyperparameters for each candidate feature set (Wrapper) approach.

I know how to do model validation + feature selection or model validation + tuning (or with filter approach), but is there any way I can perform the task above? Thank you.

**Comment by larskotthoff** [comment]
I'm not sure if I understand entirely what you're trying to do. Have you seen the part of the [tutorial](https://mlr-org.github.io/mlr-tutorial/devel/html/feature_selection/index.html) that explains how to fuse a learner with feature selection and then tune it?

Note that random forests perform feature selection internally and it is usually not necessary to do it externally.

**Comment by tjpalanca** [comment]
Apologies that I wasn't clear.

The nested-nested resampling I mentioned would ideally be as follows. (I'm using SVMs here instead of RFs as you're right about RF internal feature selection).

```
library(mlr)
library(magrittr)
makeLearner(cl = "classif.ksvm", predict.type = "prob") %>% 
  # Inner tuning loop
  makeTuneWrapper(
    par.set = makeParamSet(
      makeDiscreteParam("C", values = 2^(-2:2)),
      makeDiscreteParam("sigma", values = 2^(-2:2))
    ), 
    measures = list(auc, acc),
    control = makeTuneControlGrid(),
    resampling = makeResampleDesc(method = "CV", iters = 10)
  ) %>% 
  # Middle feature selection loop
  makeFeatSelWrapper(
    measures = list(auc, acc),
    control = makeFeatSelControlExhaustive(),
    resampling = makeResampleDesc(method = "CV", iters = 10)
  ) %>% 
  # Outer model performance estimation loop
  benchmark(
    resamplings = makeResampleDesc(method = "RepCV", reps = 10, folds = 10),
    measures = list(auc, acc)
  ) ->
  svm.model
```

So there would be two wrappers as there will be resampling -> feature selection -> tuning within each iteration. The way I'm doing this right now is to use `modelr` to do the outer model validation loop then training a one wrapper learner on each of those iterations, but it would be good to be able to do this within mlr.

Please let me know if I am approaching this correctly. Thank you!

**Comment by mb706** [comment]
[The offending lines](https://github.com/mlr-org/mlr/blob/e3f74ddccceb3d1a04f5c477d9cc1c6ce3f2e5c8/R/BaseWrapper.R#L4-L5) apparently exist because wrapping a `TuneWrapper` with something else is often a user error (note the subtle difference between having the `TuneWrapper` *before* the `FeatSelWrapper` and the other way around), there is no technical reason to prevent this as far as I can see.

**Comment by tjpalanca** [comment]
I think failing only when wrapping a `TuneWrapper` class would be the ideal solution. It would also solve my question in #1806 wrapped imputation + feature selection would fail if we select a feature set that is does not contain all variables to be computed - we can simply wrap the imputation after the feature selection so that imputation is done before subsetting.

**Comment by tjpalanca** [comment]
Closed by #1871
